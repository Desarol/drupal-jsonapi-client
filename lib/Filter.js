"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Based on: https://www.drupal.org/docs/8/modules/jsonapi/filtering
 */
var Filter =
/*#__PURE__*/
function () {
  function Filter(_ref) {
    var identifier = _ref.identifier,
        path = _ref.path,
        _ref$operator = _ref.operator,
        operator = _ref$operator === void 0 ? '=' : _ref$operator,
        value = _ref.value,
        memberOf = _ref.memberOf;

    _classCallCheck(this, Filter);

    this.identifier = identifier;
    this.path = path;
    this.operator = operator;
    this.value = value;
    this.memberOf = memberOf;
  }
  /* eslint-disable prefer-template */


  _createClass(Filter, [{
    key: "query",
    value: function query() {
      var _this = this;

      return ["filter[".concat(this.identifier, "][condition][path]=").concat(this.path), this.operator !== '=' ? "filter[".concat(this.identifier, "][condition][operator]=").concat(this.operator) : '', this.memberOf ? "filter[".concat(this.identifier, "][condition][memberOf]=").concat(this.memberOf) : '', typeof this.value.map === 'function' ? this.value.map(function (singleValue) {
        return "filter[".concat(_this.identifier, "][condition][value][]=").concat(singleValue);
      }) : "filter[".concat(this.identifier, "][condition][value]=").concat(this.value)];
    }
    /* eslint-enable prefer-template */

  }]);

  return Filter;
}();

exports.default = Filter;