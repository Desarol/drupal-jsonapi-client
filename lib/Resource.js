"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _QueryParameters = _interopRequireDefault(require("./QueryParameters"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const withMimeType = req => ({ ...req,
  headers: { ...req.headers,
    Accept: 'application/vnd.api+json',
    'Content-Type': 'application/vnd.api+json'
  }
});

class Resource {
  static Get({
    type,
    id
  }) {
    const [p1, p2] = type.split('--');
    return withMimeType({
      method: 'GET',
      url: `/${p1}/${p2}/${id}`
    });
  }

  static GetList({
    type,
    filter = null,
    include = [],
    offset = 0,
    limit = 50
  }) {
    const [p1, p2] = type.split('--');
    const params = [`page[offset]=${offset}`, `page[limit]=${limit}`];

    if (filter != null) {
      params.push(typeof filter.query === 'function' ? filter.query() : filter);
    }

    if (include.length > 0) {
      params.push(`include=${include.join(',')}`);
    }

    const queryParameters = new _QueryParameters.default(params);
    return withMimeType({
      method: 'GET',
      url: `/${p1}/${p2}?${queryParameters.toString(Number.MAX_SAFE_INTEGER)}`
    });
  }

  static New({
    type,
    data = {}
  }) {
    const [p1, p2] = type.split('--');
    return withMimeType({
      url: `/${p1}/${p2}`,
      method: 'POST',
      data: JSON.stringify({
        data: {
          type,
          ...data
        }
      })
    });
  }

  static Update({
    type,
    id,
    data = {}
  }) {
    const [p1, p2] = type.split('--');
    return withMimeType({
      url: `/${p1}/${p2}/${id}`,
      method: 'PATCH',
      data: JSON.stringify({
        data: {
          type,
          id,
          ...data
        }
      })
    });
  }

}

var _default = Resource;
exports.default = _default;