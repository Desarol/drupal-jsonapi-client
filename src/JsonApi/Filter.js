/**
 * Based on: https://www.drupal.org/docs/8/modules/jsonapi/filtering
 */

export default class Filter {
  constructor({
    identifier,
    path,
    operator = '=',
    value,
    memberOf,
  }) {
    this.identifier = identifier
    this.path = path
    this.operator = operator
    this.value = value
    this.memberOf = memberOf
  }

  /* eslint-disable prefer-template */
  query() {
    return ([
      `filter[${this.identifier}][condition][path]=${this.path}`,
      (
        this.operator !== '='
          ? `filter[${this.identifier}][condition][operator]=${this.operator}`
          : ''
      ),
      (
        this.memberOf
          ? `filter[${this.identifier}][condition][memberOf]=${this.memberOf}`
          : ''
      ),
      (
        typeof this.value.map === 'function'
          ? this.value.map(singleValue => `filter[${this.identifier}][condition][value][]=${singleValue}`)
          : `filter[${this.identifier}][condition][value]=${this.value}`
      ),
    ])
  }
  /* eslint-enable prefer-template */
}
