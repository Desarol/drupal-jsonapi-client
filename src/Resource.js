import QueryParameters from './QueryParameters'

const withMimeType = req => ({
  ...req,
  headers: {
    ...req.headers,
    Accept: 'application/vnd.api+json',
    'Content-Type': 'application/vnd.api+json',
  },
})

class Resource {
  static Get({
    type,
    id,
  }) {
    const [p1, p2] = type.split('--')

    return withMimeType({
      method: 'GET',
      url: `/${p1}/${p2}/${id}`,
    })
  }

  static GetList({
    type,
    filter = null,
    include = [],
    offset = 0,
    limit = 50,
  }) {
    const [p1, p2] = type.split('--')
    const params = [
      `page[offset]=${offset}`,
      `page[limit]=${limit}`,
    ]

    if (filter != null) {
      params.push(typeof filter.query === 'function' ? filter.query() : filter)
    }

    if (include.length > 0) {
      params.push(`include=${include.join(',')}`)
    }

    const queryParameters = new QueryParameters(params)
    return withMimeType({
      method: 'GET',
      url: `/${p1}/${p2}?${queryParameters.toString(Number.MAX_SAFE_INTEGER)}`,
    })
  }

  static New({
    type,
    data,
  }) {
    const [p1, p2] = type.split('--')

    return withMimeType({
      url: `/${p1}/${p2}`,
      method: 'POST',
      data: JSON.stringify(data),
    })
  }
}

export default Resource
