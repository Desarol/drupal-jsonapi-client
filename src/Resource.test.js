import Filter from './Filter'
import Resource from './Resource'

describe('Resource', () => {
  it('generates a valid request from a .Get call', () => {
    const request = Resource.Get({
      type: 'node--article',
      id: 'uuid-1234',
    })

    expect(request).toMatchSnapshot()
  })

  it('generates a valid request from a .GetList call', () => {
    const request = Resource.GetList({
      type: 'node--article',
    })

    expect(request).toMatchSnapshot()
  })

  it('generates a valid filters from a .GetList call', () => {
    const request = Resource.GetList({
      type: 'node--article',
      filter: new Filter({
        identifier: '1',
        path: 'uid.name',
        value: 'foo',
      }),
    })

    expect(request).toMatchSnapshot()
  })

  it('generates a valid request from a .New call', () => {
    const request = Resource.GetList({
      type: 'node--article',
      filter: new Filter({
        identifier: '1',
        path: 'uid.name',
        value: 'foo',
      }),
    })

    expect(request).toMatchSnapshot()
  })
})
