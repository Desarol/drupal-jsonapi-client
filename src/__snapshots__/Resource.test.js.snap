// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource generates a valid request from a .Get call 1`] = `
Object {
  "headers": Object {
    "Accept": "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  },
  "method": "GET",
  "url": "/node/article/uuid-1234",
}
`;

exports[`Resource generates a valid request from a .GetList call 1`] = `
Object {
  "headers": Object {
    "Accept": "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  },
  "method": "GET",
  "url": "/node/article?page[offset]=0&page[limit]=50",
}
`;

exports[`Resource generates a valid request from a .New call 1`] = `
Object {
  "data": "{\\"data\\":{\\"type\\":\\"node--article\\",\\"attributes\\":{\\"title\\":\\"title\\"}}}",
  "headers": Object {
    "Accept": "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  },
  "method": "POST",
  "url": "/node/article",
}
`;

exports[`Resource generates a valid request from a .Update call 1`] = `
Object {
  "data": "{\\"data\\":{\\"type\\":\\"node--article\\",\\"id\\":\\"uuid-1234\\",\\"attributes\\":{\\"title\\":\\"title\\"}}}",
  "headers": Object {
    "Accept": "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  },
  "method": "PATCH",
  "url": "/node/article/uuid-1234",
}
`;

exports[`Resource generates valid filters and sort from a .GetList call 1`] = `
Object {
  "headers": Object {
    "Accept": "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  },
  "method": "GET",
  "url": "/node/article?page[offset]=0&page[limit]=50&filter[1][condition][path]=uid.name&filter[1][condition][value]=foo&sort[sort-created][path]=created",
}
`;

exports[`Resource generates valid filters from a .GetList call 1`] = `
Object {
  "headers": Object {
    "Accept": "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  },
  "method": "GET",
  "url": "/node/article?page[offset]=0&page[limit]=50&filter[1][condition][path]=uid.name&filter[1][condition][value]=foo",
}
`;
